### PR TITLE
Add linear fst wrapper.

### PR DESCRIPTION
**Pull request by yaguanghu**
As mentioned in [#77](https://github.com/k2-fsa/snowfall/issues/77)

### CODE REVIEW

File: k2/python/k2/fsa_algo.py [hunk 1]
```diff
@@ -36,6 +37,33 @@ def linear_fsa(symbols: Union[List[int], List[List[int]]]) -> Fsa:
     return fsa
 
 
+def linear_fst(symbols: Union[List[int], List[List[int]]]) -> Fsa:
```
Comment (csukuangfj) [inline k2/python/k2/fsa_algo.py:40]: @danpovey 
Can you imagine in which situation is this function helpful?

```python
fsa.aux_labels = fsa.labels.clone()
```
can achieve the same goal in case it is really needed.

Comment (danpovey) [reply k2/python/k2/fsa_algo.py:40]: I agree, there is not really any point to this if the ilabels and olabels are the same, because compose() will treat it the same as a transducer with separate olabels and ilabels.

Comment (yaguanghu) [reply k2/python/k2/fsa_algo.py:40]: As #554 mentioned, if an FSA does not have aux_labels, compose is equivalent to intersect. So should we modify compose function to treat an acceptor the same as a transducer?

Comment (danpovey) [reply k2/python/k2/fsa_algo.py:40]: I believe it already does.

Comment (yaguanghu) [reply k2/python/k2/fsa_algo.py:40]: It seems not like that.  
See https://github.com/k2-fsa/k2/blob/bbeb3ea65b1bfe9aec6ba62a06d42a6c78ad19a0/k2/python/k2/fsa_algo.py#L203-L207
I don't thinks this logic will treat an acceptor the same as a transducer.

Comment (qindazhu) [reply k2/python/k2/fsa_algo.py:40]: Agree, the above API is meaningless.

I was thinking we were talking about API like below https://github.com/k2-fsa/snowfall/issues/77#issuecomment-764368954:

```
linear_fst(symbols, aux_symbols) where symbols could be a list or a list of list , aux_symbols could be a list or a ragged tensor (2 axes) from single FSA, and be a list of list or a ragged tensor (3 axes).
```
though I prefer approach `fsa.aux_labels = some_tensor/ragged_tensor` for all cases, but I'm fine with above API.

**Comment by danpovey** [comment]
I think you are mistaken about that.

File: k2/python/k2/fsa_algo.py (hunk 1 above)
Comment (yaguanghu) [reply k2/python/k2/fsa_algo.py:40]: OK, this api is more general. I will revise it.

**Comment by danpovey** [comment]
Actually you might be right.   I think perhaps these lines:
```
 if not hasattr(b_fsa, 'aux_labels'):
        return intersect(a_fsa, b_fsa, treat_epsilons_specially)
```
should be removed.

**Comment by danpovey** [comment]
Wait, no, I was wrong again, the code later on is wrong for that.
Where it says:
```
   out_fsa.labels = index(a_fsa_inv.aux_labels, a_arc_map)
    out_fsa.aux_labels = index(b_fsa.aux_labels, b_arc_map)
```
it would have to say:
```
    out_fsa.aux_labels = (index(b_fsa.aux_labels, b_arc_map) if defined b_fsa.aux_labels
                                        else out_fsa.labels)
    out_fsa.labels = index(a_fsa_inv.aux_labels, a_arc_map)

```

if inner_labels is not None:
        # out_fsa.`inner_labels` = out_fsa.labels
        setattr(out_fsa, inner_labels, out_fsa.labels)
    out_fsa.labels = index(a_fsa_inv.aux_labels, a_arc_map)
    out_fsa.aux_labels = index(b_fsa.aux_labels, b_arc_map)

**Comment by danpovey** [comment]
Wait, no, there are more problems.  I was sure the logic was right at some point.  I'll make a PR.

File: k2/python/k2/fsa_algo.py (hunk 1 above)
Comment (qindazhu) [reply k2/python/k2/fsa_algo.py:40]: sorry we can just use list of list (of list)here instead of ragged tensor, Then the input would be more pythonic

**Comment by csukuangfj** [comment]
The `aux_labels` of an FSA is usually embedded in the string form. In which case will it be separated from the string form and be passed as a list of integers to `linear_fst`?

**Comment by danpovey** [comment]
I can imagine situations where someone might have two linear sequences with the same length, corresponding to input and output.

**Comment by qindazhu** [comment]
> In which case will it be separated from the string form and be passed as a list of integers to linear_fst?

Can't python list contain string or char?

BTW, I just thought it may be helpful we can create a function that converts Python `List [of List [of List]]`  to tensor or a ragged tensor. It should be a template, i.e. works for str or int, even float.

**Comment by csukuangfj** [comment]
> Can't python list contain string or char?



A Python list can contain any Python object, including `str`.

**Comment by csukuangfj** [comment]
Could you write some tests for linear_fst ?

Thursday, 21 January 2021, 21:04 +0800 from notifications@github.com  <notifications@github.com>:

File: k2/python/k2/fsa_algo.py [hunk 2]
```diff
@@ -36,6 +37,38 @@ def linear_fsa(symbols: Union[List[int], List[List[int]]]) -> Fsa:
...
+      The scores of arcs in the returned FST are all 0.
+
+    Args:
+      symbols:
+        A list of integers or a list of list of integers.
+      aux_symbols:
+        A list of integers or a list of list of integers.
+
+    Returns:
+      An FST if the input is a list of integers.
```
Comment (qindazhu) [inline k2/python/k2/fsa_algo.py:55]: I would to say `symbols` (and `aux_symbols)` instead of `input`

File: k2/python/k2/fsa_algo.py [hunk 3]
```diff
@@ -36,6 +37,38 @@ def linear_fsa(symbols: Union[List[int], List[List[int]]]) -> Fsa:
...
+      A vector of FSTs if the input is a list of list of integers.
+    '''
+    ragged_arc = _k2.linear_fsa(symbols)
+    if isinstance(symbols[0], List):
+        assert isinstance(aux_symbols[0],
+                          List), 'aux_symbols and symbols does not match.'
+        for sym in aux_symbols:
+            sym.append(-1)  # -1 == kFinalSymbol
+        aux_labels = torch.IntTensor(aux_symbols).reshape(-1,)
```
Comment (qindazhu) [inline k2/python/k2/fsa_algo.py:64]: I wonder the behavior here: what's the size of `aux_labels`? it equals to `len(sym)` before appending or after appending? 

BTW, I guess it may be better if we also append -1 to aux_labels as well, so that when we do `invert` on the resulted linear_fst, everything would be fine naturally.

File: k2/python/k2/fsa_algo.py [hunk 4]
```diff
@@ -36,6 +37,38 @@ def linear_fsa(symbols: Union[List[int], List[List[int]]]) -> Fsa:
     return fsa
 
 
+def linear_fst(symbols: Union[List[int], List[List[int]]],
+               aux_symbols: Union[List[int], List[List[int]]]) -> Fsa:
+    '''Construct an linear FST from symbols and it's corresponding
+       auxiliary symbols.
+
```
Comment (qindazhu) [inline k2/python/k2/fsa_algo.py:44]: Maybe better to add a Caution here to say that we don't support aux_labels as a ragged tensor for now. We can fix it later if we really find we need `linear_fst` to handle this case (i.e. aux_symbols is ragged tensor)

Comment (danpovey) [reply k2/python/k2/fsa_algo.py:44]: I prefer to use `labels` instead of `symbols` here.  I would like to standardize on only using the word 'symbols' in the context of 'symbol table', and use 'labels' elsewhere.

Comment (qindazhu) [reply k2/python/k2/fsa_algo.py:44]: +2, can't agree any more.

Comment (csukuangfj) [reply k2/python/k2/fsa_algo.py:44]: This is a historical issue. 

The C++ API has been using `symbols` and `linear_fsa` just follows the convention. Later on, `symbols` in Python
means symbol table but then argument name for `k2.linear_fsa` remains up to now.

**Comment by danpovey** [comment]
ok.  we should fix.

File: k2/python/k2/fsa_algo.py (hunk 3 above)
Comment (yaguanghu) [reply k2/python/k2/fsa_algo.py:64]: The size of aux_labels is `len(sym)` after appending, because we should keep the size of aux_labels and arcs the same.
I appened -1 to aux_symbols and then converted it to torch tensor as `aux_labels`

File: k2/python/k2/fsa_algo.py (hunk 4 above)
Comment (yaguanghu) [reply k2/python/k2/fsa_algo.py:44]: I will change `symbols` to `labels` also in `linear_fsa`.

File: k2/python/k2/fsa_algo.py (hunk 3 above)
Comment (csukuangfj) [inline k2/python/k2/fsa_algo.py]: Does `torch.IntTensor` handle the case when each sublist in `aux_symbols` has a
different length?

File: k2/python/k2/fsa_algo.py [hunk 5]
```diff
@@ -17,25 +18,58 @@
...
     '''
-    ragged_arc = _k2.linear_fsa(symbols)
+    ragged_arc = _k2.linear_fsa(labels)
     fsa = Fsa(ragged_arc)
     return fsa
 
 
+def linear_fst(labels: Union[List[int], List[List[int]]],
+               aux_labels: Union[List[int], List[List[int]]]) -> Fsa:
+    '''Construct an linear FST from labels and it's corresponding
```
Comment (csukuangfj) [inline k2/python/k2/fsa_algo.py:42]: its corresponding

File: k2/python/k2/fsa_algo.py [hunk 6]
```diff
@@ -17,25 +18,58 @@
...
+       auxiliary labels.
```
Comment (csukuangfj) [inline k2/python/k2/fsa_algo.py:43]: Please remove the three extra spaces before `auxiliary labels`.

File: k2/python/k2/fsa_algo.py [hunk 7]
```diff
@@ -17,25 +18,58 @@
...
+        A list of integers or a list of list of integers.
+
+    Returns:
+      An FST if the labels is a list of integers.
+      A vector of FSTs if the input is a list of list of integers.
+    '''
+    ragged_arc = _k2.linear_fsa(labels)
+    if isinstance(labels[0], List):
+        assert isinstance(aux_labels[0],
+                          List), 'aux_labels and labels does not match.'
```
Comment (csukuangfj) [inline k2/python/k2/fsa_algo.py:61]: typo: do not match

File: k2/python/k2/fsa_algo.py (hunk 3 above)
Comment (yaguanghu) [reply k2/python/k2/fsa_algo.py]: fixed it.

Comment (danpovey) [reply k2/python/k2/fsa_algo.py]: I think it should be k2.RaggedInt, no?  
also rename all symbols->labels.

File: k2/python/k2/fsa_algo.py [hunk 8]
```diff
@@ -17,25 +18,58 @@
...
+    ragged_arc = _k2.linear_fsa(labels)
+    if isinstance(labels[0], List):
+        assert isinstance(aux_labels[0],
+                          List), 'aux_labels and labels do not match.'
+        flattened_labels = []
+        for aux in aux_labels:
+            flattened_labels.extend(aux + [-1])  # -1 == kFinalSymbol
+        aux_labels = torch.IntTensor(flattened_labels)
+    else:
+        aux_labels.append(-1)  # -1 == kFinalSymbol
```
Comment (csukuangfj) [inline k2/python/k2/fsa_algo.py:67]: This changes the input `aux_labels` **in-place**. 

I would suggest making a copy of it.

File: k2/python/k2/fsa_algo.py (hunk 3 above)
Comment (yaguanghu) [reply k2/python/k2/fsa_algo.py]: Currently `compose` function seems to support `torch.tensor` only, see
https://github.com/k2-fsa/k2/blob/bbeb3ea65b1bfe9aec6ba62a06d42a6c78ad19a0/k2/python/k2/fsa_algo.py#L209

File: k2/python/tests/linear_fst_test.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env python3
```
Comment (csukuangfj) [inline k2/python/tests/linear_fst_test.py:1]: Please also add this file to `python/tests/CMakeLists.txt`.

File: k2/python/k2/fsa_algo.py (hunk 3 above)
Comment (danpovey) [reply k2/python/k2/fsa_algo.py]: Even if torch.IntTensor handles that, which I don't think it does, our code internally does not, it expects our own ragged tensor.

Comment (danpovey) [reply k2/python/k2/fsa_algo.py]: That is specific to 'compose'.  The FSA format in general supports aux_labels being a k2.RaggedInt.

Comment (yaguanghu) [reply k2/python/k2/fsa_algo.py]: So if I use `_k2.RaggedInt`, it would not work with `compose`, is my understanding right?

Comment (csukuangfj) [reply k2/python/k2/fsa_algo.py]: > I think it should be k2.RaggedInt, no?

Currently, `RaggedInt` has only two constructors available in Python:
```
    1. _k2.RaggedInt(arg0: str)
    2. _k2.RaggedInt(shape: k2::RaggedShape, values: at::Tensor)
```

Shall we add another constructor supporting `List[List[int]]` ?

Comment (yaguanghu) [reply k2/python/k2/fsa_algo.py]: I can take this task 😊

**Comment by csukuangfj** [comment]
+1

**Comment by danpovey** [comment]
I wrote this about 24 hours ago by email, but it bounced due to rate-limit 😡.
"yes, another constructor would be good."

**Comment by yaguanghu** [comment]
I will work on it. But can we merge this change first?

**Comment by danpovey** [comment]
Is this change needed for something?

**Comment by yaguanghu** [comment]
It's not yet. I will append the revised `RaggedInt` constructor to this change.

**Comment by danpovey** [comment]
I guess it's OK to merge this.. if someone wants to.

**Comment by yaguanghu** [comment]
I expect to finish the contructor of `RaggedInt` tomorrow, please wait for that.

**Comment by csukuangfj** [comment]
> I expect to finish the contructor of RaggedInt tomorrow, please wait for that.

Thanks. Let's merge it first. You can create a new pull-request to address it.
